File: convert.js
// convert.js
const fs = require('fs');
const path = require('path');
const sharp = require('sharp');
const glob = require('glob');

async function convertHeic(inputPath, outputPath, format, quality, resizeW, resizeH) {
    let pipeline = sharp(inputPath);
    if (resizeW > 0 && resizeH > 0) {
        pipeline = pipeline.resize(resizeW, resizeH);
    }
    if (format === 'jpg' || format === 'jpeg') {
        await pipeline.jpeg({ quality }).toFile(outputPath);
    } else if (format === 'png') {
        await pipeline.png({ quality }).toFile(outputPath);
    } else if (format === 'webp') {
        await pipeline.webp({ quality }).toFile(outputPath);
    } else {
        throw new Error(`Неизвестный формат: ${format}`);
    }
}

function findFiles(inputs, recursive) {
    const exts = ['.heic', '.heif'];
    const files = [];
    for (const item of inputs) {
        if (fs.existsSync(item) && fs.statSync(item).isFile() && exts.includes(path.extname(item).toLowerCase())) {
            files.push(item);
        } else if (fs.existsSync(item) && fs.statSync(item).isDirectory()) {
            const pattern = recursive ? `${item}/**/*.*` : `${item}/*.*`;
            const matches = glob.sync(pattern, { nodir: true });
            for (const m of matches) {
                if (exts.includes(path.extname(m).toLowerCase())) {
                    files.push(m);
                }
            }
        } else if (item.includes('*')) {
            const matches = glob.sync(item);
            for (const m of matches) {
                if (exts.includes(path.extname(m).toLowerCase())) {
                    files.push(m);
                }
            }
        }
    }
    return files;
}

function processFiles(inputs, format, quality, resizeW, resizeH, outputDir, overwrite, recursive) {
    const files = findFiles(inputs, recursive);
    if (files.length === 0) {
        console.log('Не найдено HEIC-файлов.');
        return;
    }
    if (!fs.existsSync(outputDir)) {
        fs.mkdirSync(outputDir, { recursive: true });
    }
    const total = files.length;
    console.log(`Найдено ${total} HEIC-файлов.`);
    (async () => {
        for (let i=0; i<total; i++) {
            const inputFile = files[i];
            const ext = format === 'jpg' ? '.jpg' : `.${format}`;
            const outName = path.basename(inputFile, path.extname(inputFile)) + ext;
            const outPath = path.join(outputDir, outName);
            if (fs.existsSync(outPath) && !overwrite) {
                console.log(`[${i+1}/${total}] ${outPath} уже существует, пропуск.`);
                continue;
            }
            console.log(`[${i+1}/${total}] Конвертация ${inputFile} -> ${outPath}`);
            try {
                await convertHeic(inputFile, outPath, format, quality, resizeW, resizeH);
            } catch (err) {
                console.error(`  Ошибка при конвертации ${inputFile}: ${err.message}`);
            }
        }
        console.log('Готово!');
    })();
}

function main() {
    const args = process.argv.slice(2);
    if (args.length === 0) {
        console.log('Использование: node convert.js <HEIC-файлы/папки> [--format jpg|png|webp] [--quality N] [--resize ШxВ] [--output DIR] [--overwrite] [--recursive]');
        return;
    }
    let format = 'jpg';
    let quality = 85;
    let resizeW = 0, resizeH = 0;
    let outputDir = '.';
    let overwrite = false;
    let recursive = false;
    const inputs = [];
    for (let i=0; i<args.length; i++) {
        switch (args[i]) {
            case '--format':
                if (i+1 < args.length) format = args[++i];
                break;
            case '--quality':
                if (i+1 < args.length) quality = parseInt(args[++i]);
                break;
            case '--resize':
                if (i+1 < args.length) {
                    const s = args[++i];
                    const parts = s.split('x');
                    if (parts.length === 2) {
                        resizeW = parseInt(parts[0]);
                        resizeH = parseInt(parts[1]);
                    }
                }
                break;
            case '--output':
                if (i+1 < args.length) outputDir = args[++i];
                break;
            case '--overwrite':
                overwrite = true;
                break;
            case '--recursive':
                recursive = true;
                break;
            default:
                inputs.push(args[i]);
        }
    }
    processFiles(inputs, format, quality, resizeW, resizeH, outputDir, overwrite, recursive);
}

main();

File: convert.rb
# convert.rb
require 'mini_magick'
require 'find'
require 'optparse'

def convert_heic(input_path, output_path, format, quality, resize_w, resize_h)
  image = MiniMagick::Image.open(input_path)
  if resize_w > 0 && resize_h > 0
    image.resize "#{resize_w}x#{resize_h}"
  end
  image.format format
  image.quality quality if %w[jpg jpeg webp].include?(format)
  image.write output_path
end

def process_files(inputs, format, quality, resize_w, resize_h, output_dir, overwrite, recursive)
  files = []
  exts = ['.heic', '.heif']
  inputs.each do |item|
    if File.file?(item) && exts.include?(File.extname(item).downcase)
      files << item
    elsif File.directory?(item)
      if recursive
        Find.find(item) do |path|
          files << path if File.file?(path) && exts.include?(File.extname(path).downcase)
        end
      else
        Dir.glob(File.join(item, '*')).each do |path|
          files << path if File.file?(path) && exts.include?(File.extname(path).downcase)
        end
      end
    elsif item.include?('*')
      Dir.glob(item).each { |f| files << f if File.file?(f) && exts.include?(File.extname(f).downcase) }
    end
  end
  if files.empty?
    puts "Не найдено HEIC-файлов."
    return
  end
  Dir.mkdir(output_dir) unless Dir.exist?(output_dir)
  total = files.size
  puts "Найдено #{total} HEIC-файлов."
  files.each_with_index do |input_file, idx|
    ext = format == 'jpg' ? '.jpg' : ".#{format}"
    out_name = File.basename(input_file, '.*') + ext
    out_path = File.join(output_dir, out_name)
    if File.exist?(out_path) && !overwrite
      puts "[#{idx+1}/#{total}] #{out_path} уже существует, пропуск."
      next
    end
    puts "[#{idx+1}/#{total}] Конвертация #{input_file} -> #{out_path}"
    begin
      convert_heic(input_file, out_path, format, quality, resize_w, resize_h)
    rescue => e
      puts "  Ошибка при конвертации #{input_file}: #{e.message}"
    end
  end
  puts "Готово!"
end

options = {}
OptionParser.new do |opts|
  opts.banner = "Использование: ruby convert.rb <HEIC-файлы/папки> [опции]"
  opts.on("--format FORMAT", "Выходной формат (jpg, png, webp)") { |v| options[:format] = v }
  opts.on("--quality N", Integer, "Качество (1-100)") { |v| options[:quality] = v }
  opts.on("--resize ШxВ", String, "Изменение размера") { |v| options[:resize] = v }
  opts.on("--output DIR", String, "Папка для сохранения") { |v| options[:output] = v }
  opts.on("--overwrite", "Перезаписывать") { options[:overwrite] = true }
  opts.on("--recursive", "Рекурсивный обход") { options[:recursive] = true }
end.parse!

format = options[:format] || 'jpg'
quality = options[:quality] || 85
resize = options[:resize] ? options[:resize].split('x').map(&:to_i) : [0,0]
output_dir = options[:output] || '.'
overwrite = options[:overwrite] || false
recursive = options[:recursive] || false
inputs = ARGV

if inputs.empty?
  puts "Не указаны файлы или папки."
  exit
end

process_files(inputs, format, quality, resize[0], resize[1], output_dir, overwrite, recursive)
